File: Python/partial_function_application.py
#!/usr/bin/env python

from math import sqrt, exp, pi
from functools import partial

def normal_prob(
    mu  : float,
    sig : float,
    x   : float,
) -> float:
    n = 1 / (sig * sqrt(2*pi))
    z = (x-mu)/sig
    return n * exp(-(z**2)/2)

# Standard normal or Z-distribution
z_prob = partial(normal_prob, 0, 1)

def main():
    assert z_prob(1) == z_prob(-1)
    assert round(z_prob(2), 2) == 0.05

if __name__ == '__main__':
    main()
    print('SUCCESS')

File: Haskell/partial_function_application.hs
import qualified Control.Exception
assert = flip Control.Exception.assert (pure ())
--------------------------------------------------------------------------------

roundTo :: (Integral b) => Float -> b -> Float
roundTo x n = (fromIntegral . round)(x * 10^n) / 10^n

normalProb :: Float -> Float -> Float -> Float
normalProb mu sig x = n * exp(-(z**2)/2)
    where 
        n = 1 / (sig * sqrt (2*pi))
        z = (x-mu)/sig

-- Standard normal or Z-distribution
zProb = normalProb 0 1

main = do
    assert $ zProb 1 == zProb (-1)
    assert $ roundTo (zProb 2) 2 == 0.05

    putStrLn "SUCCESS"
